controllers/ratelimitpolicy_controller.go

```diff
@@ -216,7 +216,13 @@ func (r *RateLimitPolicyReconciler) deleteResources(ctx context.Context, rlp *ku
 		}
 	}
 
-	// update annotation of policies afftecting the gateway
+	// remove direct back ref from limitador CR
+	err = r.deleteLimitadorBackReference(ctx, rlp)
```

Comment:
Hmm. So `deleteLimits` → `reconcileLimitador` → `reconcileLimitadorBackRef` adds the RLP ref to the back-ref annotation and performs an update on the `Limitador` resource, then we call `deleteLimitadorBackReference`, which removes the same RLP ref from the annotation and updates the `Limitador` resource again?

Reply:
Yes, that is correct and I should explain my reasoning. The `reconcileLimitador` function runs no matter what action is being carried out let it be create, update or delete. The label removal logic could be in the `reconcileLimitadorBackRef` but doing this would mean having recreate the logic for checking if the resource has being deleted. As it is the `deleteResources` --> `deleteLimitadorBackReference` only get called when a RLP is deleted there was no need to any more checks in the `deleteLimitadorBackReference`. I am aware there are more api calls with what I have done but as its only on deletion and not every reconcile I felt the trade off was worth it. 

Do you want me to move the label deletion logic to the `reconcileLimitadorBackRef`?

Reply:
> Do you want me to move the label deletion logic to the `reconcileLimitadorBackRef`?

I would wait. As I said in my other [comment](https://github.com/Kuadrant/kuadrant-operator/pull/240#pullrequestreview-1598759818), if we're going to watch and reconcile changes on all derived resources – the `Limitador` CR is just one of then! –, I see more of this pattern of annotating/de-annotating ahead. I think we can aim for more generic functions and types that we can reuse in more places.

Reply:
From the community call my understanding is we wont be using the  https://github.com/Kuadrant/gateway-api-machinery till after MVP. So I thinking I address the other issue raised now and create an issue to follow up with a refactor of the labelling once we can use the gateway-api-machinery. Would you be happy with this approach?

Reply:
This won't be complete anyway until we cover for all kinds of resources that, once modified directly, can break with the source of truth. I understand that whereas one is a refactoring, while the other prevents what could be perceived as a bug, I still think those two things are intimately related and better if tackled together. I'm also afraid of the explosion of annotations while [others seem to be addressing the issue differently](https://github.com/kubernetes-sigs/gateway-api/pull/2128). IOW, let's not amend it quickly due to release pressure and do this the right way. The fix can be part of a post-release patch if we want.